### PR TITLE
show Passthrough cover on track overview

### PR DESCRIPTION
**Pull request by ronso0**
image

Inspired by #2575 

Show passthrough cover over entire overview, picks color from `<PassthroughOverlayColor>` node.
Stack _Passthrough_ label on top, accepts custom styles via qss: `WOverview #PassthroughLabel`.
All mouse events are still passed to the overview.

I added the `PassthroughOverlayColor` picker to /waveform/renderers/waveformsignalcolors.cpp, I guess it can also be used for the scrolling waveform then?

- [x] add styles to all skins
- [x] add default color to C++?

### CODE REVIEW

**Comment by ronso0** [comment]
It would be nice if someone colud review this to let me know if I'm on the right track.
If so, I'll continue to add the skin styles and a default color.

**Review by Holzhaus — COMMENTED** [review]
LGTM so far, thanks!

**Comment by ronso0** [comment]
well, that was quick!

**Comment by ronso0** [comment]
> Be-ing removed the beta non-blocker label

why? "passthrough" is displayed over the overview #2575 , this upgrade just keeps the overview usable and can be added during beta (if I don't finish the skin implementation before that)

**Comment by Be-ing** [comment]
"beta non-blocker" is a useless tag. Everything not tagged as a beta blocker is a non-blocker; having a tag for that is superfluous.

**Comment by ronso0** [comment]
There where some embarassing copy/paste mistakes and a bug with skin reload that slipped throught earlier, so I started over and rebased. Please review.
Some functions are also more effective now, for example the cover color is only read once during setup. I'll use that for the playedOverlayColor, too.

With default.qss it looks like this:
image

Custom skin styles coming soon..

File: src/widget/woverview.cpp [hunk 1]
```diff
@@ -94,6 +93,17 @@ WOverview::WOverview(
             this, &WOverview::onTrackAnalyzerProgress);
 
     connect(m_pCueMenuPopup.get(), &WCueMenuPopup::aboutToHide, this, &WOverview::slotCueMenuPopupAboutToHide);
+
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft|Qt::AlignVCenter);
+    m_pPassthroughLabel->setText("Passthrough");
```
Comment (Be-ing) [inline src/widget/woverview.cpp:100]:
```suggestion
    //: Shown on the overview waveform when vinyl passthrough is enabled
    m_pPassthroughLabel->setText(tr("Passthrough"));
```

**Comment by ronso0** [comment]
image
image

image
image
image

Tango with #2619 : no passthrough cover on Play, no highlight on Vinyl Controls toggle
image
image

image

File: src/widget/woverview.cpp (hunk 1 above)
Comment (ronso0) [reply src/widget/woverview.cpp]: okay, thx!

File: src/widget/woverview.h [hunk 2]
```diff
@@ -170,6 +171,9 @@ class WOverview : public WWidget, public TrackDropTarget {
     QColor m_labelTextColor;
     QColor m_labelBackgroundColor;
     QColor m_endOfTrackColor;
+    QColor m_passthroughOverlayColor;
+    QLabel* m_pPassthroughLabel;
+    QVBoxLayout* m_pPassThroughLayout;
```
Comment (Holzhaus) [inline src/widget/woverview.h:176]: `setLayout` transfers ownership to the widget, you don't need to save a reference to the layout.

File: src/widget/woverview.cpp [hunk 3]
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
             this, &WOverview::onTrackAnalyzerProgress);
 
     connect(m_pCueMenuPopup.get(), &WCueMenuPopup::aboutToHide, this, &WOverview::slotCueMenuPopupAboutToHide);
+
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft|Qt::AlignVCenter);
```
Comment (Holzhaus) [inline src/widget/woverview.cpp:99]:
```suggestion
    m_pPassthroughLabel->setAlignment(Qt::AlignLeft | Qt::AlignVCenter);
```

**Review by Holzhaus — CHANGES_REQUESTED** [review]
Great! Code looks good.

Now for the obligatory bikeshedding:
I think we should try to re-use the skin's track title label text colors for the Passthrough label. Different colors look a bit out of place. For shade, this would be dark grey instead of green, and in Tango blue or Fuchsia depending on color scheme. I'm not sure about Deere, but I think either white or blue (the same shade that is used for buttons) would look good.

File: src/widget/woverview.h (hunk 2 above)
Comment (ronso0) [reply src/widget/woverview.h:176]: so `QVBoxLayout passThroughLayout = new ...QVBoxLayout()` would suffice?

Comment (Holzhaus) [reply src/widget/woverview.h:176]:
```c++
QVBoxLayout* pPassThroughLayout = new QVBoxLayout();
```

**Comment by ronso0** [comment]
> Now for the obligatory bikeshedding:
> I think we should try to re-use the skin's track title label text colors for the Passthrough label. Different colors look a bit out of place.

Here we go :)
I think we should emphasize the _indicator_ nature of the Passthrough label because it's a different information level than a _track title_, so I chose scheme-specific indiciator/button colors*. IMO that helps to notice it, especially since the Passthrough toggles are either very small (Shade), can be hidden (Tango, Deere) or might be far away from the decks (Deere w/ stacked waveforms).
*only in LateNight this doesn't make difference

File: src/widget/woverview.h (hunk 2 above)
Comment (ronso0) [reply src/widget/woverview.h]: a got it..

File: src/widget/woverview.cpp (hunk 3 above)
Comment (ronso0) [reply src/widget/woverview.cpp]: ok

**Comment by Holzhaus** [comment]
I converted this to draft to prevent premature merging because there are still things left to do (according to the checkboxes in your PR description). If this is ready, feel free to convert it back.

File: src/widget/woverview.cpp [hunk 4]
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
...
 
     connect(m_pCueMenuPopup.get(), &WCueMenuPopup::aboutToHide, this, &WOverview::slotCueMenuPopupAboutToHide);
+
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft | Qt::AlignVCenter);
+    // Shown on the overview waveform when vinyl passthrough is enabled
+    m_pPassthroughLabel->setText(tr("Passthrough"));
+    m_pPassthroughLabel->hide();
+    QVBoxLayout *passThroughLayout = new QVBoxLayout(this);
```
Comment (Holzhaus) [inline src/widget/woverview.cpp:103]:
```suggestion
    QVBoxLayout *pPassthroughLayout = new QVBoxLayout(this);
```

**Review by Holzhaus — APPROVED** [review]
Code LGTM so far (except for one variable name). Thank you.

**Comment by ronso0** [comment]
Thx, comments addressed, ToDos finished.
Ready for merge if the colors are reasonable.

**Comment by ronso0** [comment]
> I converted this to draft

Ha! Now I found that 'button', too.

**Review by Holzhaus — APPROVED** [review]
> Ready for merge if the colors are reasonable.

I still don't like the colors (except for LateNight), but this subjective and I don't want to start endless discussions about this. I you think they fit in nicely, so be it (I'm using LateNight anyway :wink:).

File: src/widget/woverview.cpp [hunk 5]
```diff
@@ -94,6 +93,18 @@ WOverview::WOverview(
...
     connect(m_pCueMenuPopup.get(), &WCueMenuPopup::aboutToHide, this, &WOverview::slotCueMenuPopupAboutToHide);
+
+    m_pPassthroughLabel = new QLabel(this);
+    m_pPassthroughLabel->setObjectName("PassthroughLabel");
+    m_pPassthroughLabel->setAlignment(Qt::AlignLeft | Qt::AlignVCenter);
+    // Shown on the overview waveform when vinyl passthrough is enabled
+    m_pPassthroughLabel->setText(tr("Passthrough"));
+    m_pPassthroughLabel->hide();
+    QVBoxLayout *pPassthroughLayout = new QVBoxLayout(this);
+    passThroughLayout->setContentsMargins(0,0,0,0);
```
Comment (Holzhaus) [inline src/widget/woverview.cpp:104]: You only changed the first occurrence.

```suggestion
    pPassthroughLayout->setContentsMargins(0,0,0,0);
```
and below

**Review by Holzhaus — CHANGES_REQUESTED** [review]

File: src/widget/woverview.cpp (hunk 5 above)
Comment (Holzhaus) [reply src/widget/woverview.cpp:104]: Please fixup the broken commits in this branch and force-push to allow us to git bisect later.

Comment (ronso0) [reply src/widget/woverview.cpp]: Thanks for checking! Done.

**Comment by ronso0** [comment]
Fixed and squashed.
Ready for merge.

**Review by Holzhaus — COMMENTED** [review]
Thanks!

 I retested. The font for Tango is incorrect on my machine:

2020-04-15-105714_747x385_scrot

Other than that, LGTM.

**Comment by ronso0** [comment]
> I retested. The font for Tango is incorrect on my machine:

Strange, here it looks as intended.
I added `font-weight: bold;` to all stylesheets to ensure it on all OS.
After further testing I made the label color in Shade Dark a bit brighter.
For the sake of completeness I also removed the Passthrough hints from Tango. (might create conflicts for #2619 though)
Ready!

**Review by Holzhaus — COMMENTED** [review]
Still looks like this:

2020-04-15-115852_664x293_scrot

**Comment by ronso0** [comment]
This is deck1 with ClubTwist, right?
I have no idea why the font style is not applied, ClubTwist doesn't have a separate stylesheet.
Issue should be solved after merging #2647

**Comment by Holzhaus** [comment]
> This is deck1 with ClubTwist, right?
> I have no idea why the font style is not applied, ClubTwist doesn't have a separate stylesheet.
> Issue should be solved after merging #2647

I merged #2647 locally and the issue is indeed fixed.

**Review by Holzhaus — APPROVED** [review]
LGTM.

**Pull request merged**